### DISCUSSION TITLE
🐞 Setting `--oci-max-parallelism 2` causes deadlocks

### DISCUSSION THREAD

**Discussion by gerhard** [dagger/dagger#12340] [Help]
### What is the issue?

Setting `--oci-max-parallelism 2` on the Engine causes deadlocks, both with & without modules.

This is what that looks like from the CLI / user perspective:

https://github.com/dagger/dagger/assets/3342/826ddb73-c36e-469e-9560-f21b3fd13388

Attaching Dagger Engine v0.10.2 logs + `SIGQUIT`. Thank you @vito for https://vito.github.io/swirly/ and the tips!
- [dagger-engine-010x.sigquit.txt](https://github.com/dagger/dagger/files/14624908/dagger-engine-010x.sigquit.txt)
- [dagger-engine-010x.log.txt](https://github.com/dagger/dagger/files/14624909/dagger-engine-010x.log.txt)

FWIW:

https://github.com/dagger/dagger/assets/3342/43657975-5e84-41f4-b199-14bb7773a5a9

cc @sipsma @vito @jedevc @aluzzardi

### Dagger version

Confirmed on 0.9.11 & 0.10.2

**Comment by Scalahansolo** [comment]
The team over here at Motion have seen this happen to us as well on 0.10.3

**Comment by jedevc** [comment]
I think this is expected - suppose the following module structure:

- Module A starts (running a container)
- Module A calls a function in Module B
- Module B start (running a container)
- Module B calls `dag.Container()` (running a container)
	- This hangs forever! There are already 2 containers running in parallel, so no more can be scheduled.

I think with modules, this effectively means that `max-parallelism` is pretty much just completely broken - it just fits very badly with the idea of nested containers.

Not sure what the right fix is here:
- Don't use `max-parallelism` (generally, my preference, you should enforce max usage using different mechanisms instead, since a single job-step could use anywhere from 0% to 100% of available resources - they're not just replaceable)
- When using a nested step, don't have this "consume a resource" (this prevents deadlocks, but isn't great, since there's now more containers running than the user asked for)

I don't think there's a clear way forwards here, but to me this kind of highlights that we need a better way of constraining resources - `max-parallelism` is just really not fit for the shape that pipelines tend to take these days. Maybe we should provide some cgroup config options instead (if the end desire is to constrain CPU usage/etc).

**Comment by jedevc** [comment]
Adding this to the v0.12.0 milestone since by that release we should probably either:
- Remove this flag
- Ensure that nested execs don't count towards max-parallelism
As is, with modules, this flag can cause *huge* performance issues, so should be avoided.

**Comment by jedevc** [comment]
x-linking to https://github.com/dagger/dagger/issues/6465 as well.

**Comment by jedevc** [comment]
Bleh, shifting this one out of the milestone - I think we've got some rough consensus to take option 2 in https://github.com/dagger/dagger/issues/6894#issuecomment-2025654478, which is actually a backwards compatible change, so we can work on this at another point.
